### PR TITLE
FIX: training will not hang if in task an error occurs

### PR DESCRIPTION
**Pull request by razmikmelikbekyan**
*Issue #266, if available:*

*Description of changes:*
Checking queue size before calling fetch. When we call fetch and the process which is inside the queue fails, then we have deadlock: process is raising exception, put the thread which opened this process is not raising an exception.
Basic PEP8 corrections.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-313-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-313/2/index.html

File: autogluon/core/task.py [hunk 1]
```diff
@@ -40,17 +42,20 @@ def __init__(self, fn, args, resources):
 
     @classmethod
     def set_id(cls, taskid):
-        logger.info('Seting TASK ID: {}'.format(taskid))
+        logger.info(f'Seting TASK ID: {taskid}')
```
Comment (Innixma) [inline autogluon/core/task.py]: Super minor, but I noticed old typo here 'Seting' -> 'Setting'

**Comment by zhanghang1989** [comment]
This PR mainly addressed some coding style and looks good to me.

Why it fix training hang?

File: autogluon/scheduler/reporter.py [hunk 2]
```diff
@@ -64,6 +64,9 @@ def __call__(self, **kwargs):
             raise AutoGluonEarlyStop('Stopping!')
 
     def fetch(self, block=True):
+        if not self._queue.qsize():
```
Comment (jwmueller) [inline autogluon/scheduler/reporter.py]: @zhanghang1989 
It appears this is the key part of the fix, are you able to verify this is safe?

Comment (zhanghang1989) [reply autogluon/scheduler/reporter.py]: Thanks @jwmueller  for reminding. This change is not safe. Basically it breaks all the search algorithms except for random search. Because we need the blocking call of `self._queue.get()` to wait for next trial finishes.

**Comment by zhanghang1989** [comment]
Beyond the breaking of blocking call of `reporter.fetch()`, this PR looks okay for me.

@mseeger could you also review this PR and see if it breaks the backend?

File: autogluon/scheduler/fifo.py [hunk 3]
```diff
@@ -257,13 +258,17 @@ def _run_reporter(self, task, task_job, reporter, searcher):
         last_result = None
         while not task_job.done():
             reported_result = reporter.fetch()
+            if reported_result is None:
```
Comment (razmikmelikbekyan) [inline autogluon/scheduler/fifo.py]: @zhanghang1989 this is the while loop, which will continue calling `queue.get` untill it will get something from there or `task_job.done() is True`

File: autogluon/scheduler/reporter.py (hunk 2 above)
Comment (razmikmelikbekyan) [reply autogluon/scheduler/reporter.py]: @zhanghang1989 thanks for review. 

As far as I understand the logic is the following: for each training task we have a thread which itself opens a process, please correct we if I am wrong. Then we collect data in Queue from child process and pass to main process. So the issue occurs when we try to call `queue.get`, but the child process is already died due to some reason (for example user passed wrong parameters to particular model). At that moment we have a deadlock.  I assume in order to solve it, there will be a need to change some logic in the multiprocessing / multithreading part of the code. With this change I just avoid calling `queue.get` in order to not have the deadlock. And it is not blocking algorithms, since it will try to call `queue.get` until it will get something from `queue` or the job is done.. If you look at changes in searchers, you will see that it is in `while` loop. I checked it for `skopt` and it works.

@Innixma please see the minimal code example, which will cause deadlock, without these changes. 

Tabular Prediction: 

```
import autogluon as ag
from autogluon.task import TabularPrediction as task

import sklearn.datasets
import pandas as pd

cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
data = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
data['target'] = cancer_data['target']

hp_tune = True
cat_options = {
    'num_leaves': ag.space.Int(lower=26, upper=66, default=36),
    'depth': ag.space.Int(lower=4, upper=10, default=6),
}
gbm_options = {
    'num_boost_round': 100,
    'num_leaves': ag.space.Int(lower=26, upper=66, default=36),
}

hyperparameters = {
    'CAT': cat_options,
    'GBM': gbm_options,
}

time_limits = 2 * 60
num_trials = 3
output_directory = './agModels--tuningCancerDataset'
data = task.Dataset(data)
label_column = 'target'

# case 1: random search
search_strategy = 'random'
print('\n' * 2)
print('#' * 100)
print(f'Case 1: Running HPO suing search strategy = "{search_strategy}".')
print()
predictor = task.fit(
    train_data=data,
    holdout_frac=0.3,
    label=label_column,
    output_directory=output_directory,
    time_limits=time_limits,
    num_trials=num_trials,
    hyperparameter_tune=hp_tune,
    hyperparameters=hyperparameters,
    search_strategy=search_strategy,
    auto_stack=False,
    stack_ensemble_levels=0
)

print('\n' * 2)
print('#' * 100)
print('HPO results:')
print(predictor.leaderboard(silent=True))

# case 2: skopt search
search_strategy = 'skopt'
print('\n' * 2)
print('#' * 100)
print(f'Case 2: Running HPO suing search strategy = "{search_strategy}".')
print()
predictor = task.fit(
    train_data=data,
    holdout_frac=0.3,
    label=label_column,
    output_directory=output_directory,
    time_limits=time_limits,
    num_trials=num_trials,
    hyperparameter_tune=hp_tune,
    hyperparameters=hyperparameters,
    search_strategy=search_strategy,
    auto_stack=False,
    stack_ensemble_levels=0
)
print('\n' * 2)
print('#' * 100)
print('HPO results:')
print(predictor.leaderboard(silent=True))
```
Here I put wrong parameter `num_leaves` for `catboost`: `catboost` does not support changing it when you run on CPU.

Image Classification: 
```
import autogluon as ag
from autogluon import ImageClassification as task

train_data = task.Dataset(name='fashionmnist')

classifier = task.fit(train_data,
                      num_trials=1,
                      net=None,
                      epochs=5,
                      ngpus_per_trial=1,
                      verbose=False,
                      auto_search=False)
```
Here I put wrong parameter `net=None` .

**Comment by szha** [comment]
Job PR-313-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-313/3/index.html

**Comment by mseeger** [comment]
Hello @zhanghang1989, thanks for looping me in.
I am not fully on top of the implementation details of the backend.

First, I'd ask the author to split cosmetic changes into a separate PR, and focus this PR on the essential changes in question. The author should also comment on @zhanghang1989 objections.

File: autogluon/scheduler/fifo.py (hunk 3 above)
Comment (mseeger) [reply autogluon/scheduler/fifo.py]: @zhanghang1989 It comes down to whether a call of reporter.fetch() is blocking, in the sense it waits until fetch() returns something. I kind assumed this was the case.

If the call is not blocking, our current code is broken, because it would try to add a signal from the reporter even if fetch() returned an empty dict (I think this is the issue, not so much None).

Comment (mseeger) [reply autogluon/scheduler/fifo.py]: The reason I assumed that reporter.fetch() is blocking is that I see reporter.move_on() afterwards. If the reporter would simply just continue, why we would need that call?

File: autogluon/scheduler/hyperband.py [hunk 4]
```diff
@@ -219,6 +218,9 @@ def _run_reporter(self, task, task_job, reporter, searcher, terminator):
         last_updated = None
         while not task_job.done():
             reported_result = reporter.fetch()
+            if reported_result is None:
```
Comment (mseeger) [inline autogluon/scheduler/hyperband.py]: Same issues as above. I do think reporter.fetch() would not return None. The question is whether it would return an empty dict.

File: autogluon/scheduler/reporter.py (hunk 2 above)
Comment (mseeger) [reply autogluon/scheduler/reporter.py]: OK, I see now. @zhanghang1989 

I think this IS a problem. With this insertion, a call to fetch() becomes non-blocking, right? Before, it seems that fetch() would not return anything (and wait) until somethere is there to return (please confirm).

But now, it immediately returns None if nothing is there to report, and this is not what we want. This would really break things for the schedulers.

To be clear, there are three different things that could happen when call fetch():
A) Nothing new to be reported
B) Something new to be reported, but not a final metric at the end of training
C) Something new to be reported, namely final metric

Currently, schedulers assume that for A), reporter.fetch() blocks, until B) or C). And scheduler then has to distinguish between them.

Is this blocking the scheduler from doing other things? No, because we have a reporter thread, and only that is blocked (it runs _run_reporter)

Comment (razmikmelikbekyan) [reply autogluon/scheduler/reporter.py]: What if the process is killed, which is supposed to return something in queue ? Then we have deadlock.

**Comment by mseeger** [comment]
I only have a bit of a high-level view on the workings here.
Please do discuss this with @zhanghang1989 

Maybe it is the case that changing fetch() to return None sometimes, and then insert continue if None is received in _run_reporter of several different schedulers keeps everything working, and on top serves a purpose I do not understand. If so, we should change it. For now, I do not see why reporter.fetch() should not wait until something can be returned. This may be my ignorance.

File: autogluon/scheduler/fifo.py (hunk 3 above)
Comment (zhanghang1989) [reply autogluon/scheduler/fifo.py]: @mseeger Thanks for reviewing the PR.` reporter.fetch()` is a blocking call. This PR does break our code.

**Comment by Innixma** [comment]
@zhanghang1989 Could you produce an example where this change breaks functionality that previously worked, as Razmik shows an example where his change has a positive impact on Tabular where previously it resulted in deadlock.

**Comment by zhanghang1989** [comment]
> @zhanghang1989 Could you produce an example where this change breaks functionality that previously worked, as Razmik shows an example where his change has a positive impact on Tabular where previously it resulted in deadlock.

Basically the searcher state won't get updated in real example, because the results are not ready in reporter fetch. Only random searcher works as expected.

**Comment by Innixma** [comment]
I understand the claim, but can you provide a code example of this not working as intended such that we can work towards a solution where both your example and Razmik's example work correctly?

**Comment by zhanghang1989** [comment]
The easiest way to handle dead trials is adding a time_out, which I can help with.

**Comment by Innixma** [comment]
We shouldn't be doing that, they aren't dead trials, they throw an exception which can be observed. We want to avoid the deadlock entirely.

**Comment by mseeger** [comment]
Hello,

here is my view, which may be inaccurate:

I think the change does not break our code, because two things happen:
- reporter.fetch() returns None if nothing is to be reported, so the call becomes non-blocking
- But then, _run_reporter in the scheduler do continue if the result is None, so the blocking is
   simply moved to the _run_reporter loop

My real question would be why this is useful, because it clearly makes _run_reporter implementations a little more complex: we always have to remember to react to None in the right way.

@razmikmelikbekyan Could you explain to us (best with an example) why the added complexity actually helps over the previous code, where reporter.fetch() was blocking? I am not experienced in distributed systems.

**Comment by razmikmelikbekyan** [comment]
> Hello,
> 
> here is my view, which may be inaccurate:
> 
> I think the change does not break our code, because two things happen:
> 
> * reporter.fetch() returns None if nothing is to be reported, so the call becomes non-blocking
> * But then, _run_reporter in the scheduler do continue if the result is None, so the blocking is
>   simply moved to the _run_reporter loop
> 
> My real question would be why this is useful, because it clearly makes _run_reporter implementations a little more complex: we always have to remember to react to None in the right way.
> 
> @razmikmelikbekyan Could you explain to us (best with an example) why the added complexity actually helps over the previous code, where reporter.fetch() was blocking? I am not experienced in distributed systems.

@Innixma , @mseeger , @zhanghang1989 
Please refer to PR #322
